client/chat.coffee

```diff
@@ -50,11 +50,42 @@ messagesForPage = (p, opts={}) ->
     timestamp: cond
   , opts
 
+isFollowup = (former, latter) ->
+  return false unless former?.classList?.contains("media")
+  return false unless latter.classList.contains("media")
+  return false unless former.dataset.nick == latter.dataset.nick
+  if former.classList.contains("bb-message-pm")
+    return false unless latter.classList.contains("bb-message-pm")
+    return false unless former.dataset.pmTo == latter.dataset.pmTo
+  return true
+
+effectFollowup = (former, latter) ->
+  console.log former, latter
+  return unless latter?.classList?
+  if isFollowup(former, latter)
+    console.log "#{latter} is followup of #{former}"
+    latter.classList.add("bb-message-followup")
+  else
+    console.log "#{latter} is not followup of #{former}"
+    latter.classList.remove("bb-message-followup")
+
 # Globals
 instachat = {}
 instachat["UTCOffset"] = new Date().getTimezoneOffset() * 60000
 instachat["alertWhenUnreadMessages"] = false
 instachat["scrolledToBottom"]        = true
+instachat["mutationObserver"] = new MutationObserver (recs, obs) ->
+  for rec in recs
+    someElement = false
+    check = (e) -> someElement = true if e instanceof Element
+    check node for node in rec.addedNodes
+    check node for node in rec.removedNodes
+    continue unless someElement
+    prevEl = rec.previousSibling?.nextSibling?.previousElementSibling
```

Comment:
I think this should be something like:
```
prevEl = rec.previousSibling
unless prevEl instanceof Element
  prevEl = prevEl?.previousElementSibling
```
and in the case where first elements in the list are deleted, leaving a new first element, `previousSibling` will be null, and you want to check nextSibling, too:
```
nextEl = rec.nextSibling
unless nextEl instanceof Element
  nextEl = nextEl?.nextElementSibling

effectFollowup(prevEl, prevEl?.nextElementSibling)
effectFollowup(nextEl?.previousElementSibling, nextEl)
```
In fact, I *think* the `prevEl, prevEl?.nextElementSibling` case doesn't do anything useful, since in the case of added nodes, the nextElementSibling will be handled by the iteration over `addedNodes` (which should really be `effectFollowup(node.previousElementSibling, node)`).

Reply:
Re `rec.previousSibling?.nextSibling?.previousElementSibling`, I assumed following three pointers was more efficient than an instanceof check, especially when the check was probably already done by the DOM to generate the pointers.
Re `prevEl, prevEl?.nextElementSibling`, I need a case that hangs over one side or the other of addedNodes to cover elements being deleted and none added. I suppose having that be the other direction is fine, since there's always an element at the end (the text box).